pwiz_tools/Skyline/Model/Proteome/ProteinGroupMetadata.cs

```diff
@@ -0,0 +1,216 @@
+            return new ProteinGroupMetadata(this);
+        }
+
+        public override ProteinMetadata ChangeGene(string gene)
+        {
+            return new ProteinGroupMetadata(this);
+        }
+
+        public override ProteinMetadata ChangeSpecies(string species)
+        {
+            return new ProteinGroupMetadata(this);
+        }
+
+        public ProteinMetadata ChangeSingleProteinMetadata(ProteinMetadata singleProteinMetadata)
+        {
+            Assume.IsTrue(singleProteinMetadata != null && !(singleProteinMetadata is ProteinGroupMetadata));
+            var proteinMetadataList = new List<ProteinMetadata>(ProteinMetadata);
+            var matchingProteinIndex = proteinMetadataList.IndexOf(m => m.Name == singleProteinMetadata?.Name);
+            Assume.IsTrue(matchingProteinIndex >= 0, $@"no matching protein in group {this} for protein {singleProteinMetadata}");
+            proteinMetadataList[matchingProteinIndex] = singleProteinMetadata;
+            return new ProteinGroupMetadata(Name, Description, proteinMetadataList);
+        }
+
+        public override ProteinMetadata ChangeWebSearchInfo(WebSearchInfo webSearchInfo)
+        {
+            return new ProteinGroupMetadata(this, webSearchInfo);
+        }
+
+        public override ProteinMetadata ClearWebSearchInfo()
+        {
+            // sometimes all you really want is to initialize
+            return new ProteinGroupMetadata(this, WebSearchInfo.EMPTY);
+        }
+
+        public override ProteinMetadata SetWebSearchCompleted()
+        {
+            return new ProteinGroupMetadata(this, WebSearchInfo.SetSearchCompleted());
+        }
+
+        public override ProteinMetadata SetWebSearchTerm(WebSearchTerm search)
+        {
+            return new ProteinGroupMetadata(this, WebSearchInfo.SetSearchTerm(search));
+        }
+
+        /// <summary>
+        /// returns a copy of this, with any blanks filled in
+        /// using the members of source
+        /// </summary>
+        /// <param name="source">the source to merge from</param>
+        /// <returns>a copy of this merged with source, with this winning when fields conflict</returns>
+        public override ProteinMetadata Merge(ProteinMetadata source)
+        {
+            if (source == null)
+                return this;
+            if (source is ProteinGroupMetadata sourceGroup)
+            {
+                return new ProteinGroupMetadata(String.IsNullOrEmpty(Name) ? sourceGroup.Name : Name,
+                    String.IsNullOrEmpty(Description) ? sourceGroup.Description : Description,
+                    sourceGroup.ProteinMetadata);
+            }
+
+            return new ProteinGroupMetadata(String.IsNullOrEmpty(Name) ? source.Name : Name,
+                String.IsNullOrEmpty(Description) ? source.Description : Description,
+                ProteinMetadata);
+            //throw new InvalidOperationException("cannot merge ProteinMetadata into ProteinGroupMetadata");
```

Comment:
Seems unnecessary.

Reply:
The commented out exception?

Reply:
Right. Didn't realize this wouldn't be clearly attached to the line I added the comment on.

Reply:
You could use Assume.Fail() to make it clearer (and then return null - which will never happen but ReSharper doesn't know that). Programming error type messages which are expected to end up as an unhandled exception and reported to us through the unhandled exception form do not need to be localized. So, you can just make them constant strings.